DOC-2823-execution time inputs (#1684)
harness/developer-hub: docs/platform/20_References/runtime-inputs.md

Later, if you choose to rerun a Pipeline, the Pipeline will run using the Runtime Inputs you provided the last time it ran.

#### Use runtime inputs in a stage or pipeline

Using Runtime Inputs templates some or all of a stage or Pipeline's settings. The same Pipeline can be run using different values for all of the Runtime Inputs.

#### CI example

You can use Runtime Inputs in a CI stage's Infrastructure. Here's an example using a Runtime Input in the **Namespace** setting.

![](./static/runtime-inputs-05.png)

#### CD example

You can use Runtime Inputs for the Service in a CD stage's Service settings.

![](./static/runtime-inputs-06.png)

### Using runtime inputs during execution

> Currently, this feature is behind the feature flag `NG_EXECUTION_INPUT`. Contact [Harness Support](mailto:support@harness.io) to enable the feature. You can add runtime input to a pipeline that runs when a stage or a step is executed. If a custom stage is setup for runtime input, you can enter a shell script when prompted by Harness during execution. If a Harness Approval step is setup for runtime input, when the pipeline executes, you can specify the Harness Groups that will approve that step.

Reviewer: I think after this `You can add runtime input to a pipeline that runs when a stage or a step is executed.` we can show how users can make a field execution-time-input. <+input>.executionInput()
Then we can give the example. Something like
`Let say we have a custom stage with a shell-script step and the Shell-Script step has a field as execution-time-input. So on executing the pipeline, user will be prompted for the runtime input just before starting the Shell-script. `
And we can add one more example like Harness Approval step.

Later, if you choose to rerun a Pipeline, the Pipeline will run using the Runtime Inputs you provided the last time it ran.

#### Use runtime inputs in a stage or pipeline

Using Runtime Inputs templates some or all of a stage or Pipeline's settings. The same Pipeline can be run using different values for all of the Runtime Inputs.

#### CI example

You can use Runtime Inputs in a CI stage's Infrastructure. Here's an example using a Runtime Input in the **Namespace** setting.

![](./static/runtime-inputs-05.png)

#### CD example

You can use Runtime Inputs for the Service in a CD stage's Service settings.

![](./static/runtime-inputs-06.png)

### Using runtime inputs during execution

Currently, this feature is behind the feature flag `NG_EXECUTION_INPUT`. Contact [Harness Support](mailto:support@harness.io) to enable the feature. 

You can add runtime input to a pipeline that runs when a stage or a step is executed. For example, for a custom stage with a Shell Script step that has an runtime input field, you will be prompted to enter the input during the execution just before starting the Shell Script step.

> If a custom stage is setup for runtime input, you can enter a shell script when prompted by Harness during execution. 

Reviewer: I think this line is redundant to the above line and we can remove this line.

Author: The above line is already removed. Please check preview.

If a Harness Approval step is setup for runtime input, when the pipeline executes, you can specify the Harness Groups that will approve that step.

You can select the following runtime input types for a step or stage during an execution:  

* Allowed values (`allowedValues()`)
* Default value (optional) (`default()`)

> * Request input value when step/stage is being executed (`executionInput()`)

Author: Please add examples for all

> * Request input value when step/stage is being executed (`executionInput()`)

Reviewer: <+input>.allowedValues(value1,value2).executionInput()
Also explain this that when the user is prompted to provide the input for this field during the execution. Only value1 and value2 will be allowed as valid input.

<+input>.allowedValues(value1,value2).default(value1).executionInput()
Also explain this that when the user is prompted to provide the input for this field during the execution. Only value1 and value2 will be allowed as valid input. And value1 is the default value so it will come as default input in the prompt.

Later, if you choose to rerun a Pipeline, the Pipeline will run using the Runtime Inputs you provided the last time it ran.

#### Use runtime inputs in a stage or pipeline

Using Runtime Inputs templates some or all of a stage or Pipeline's settings. The same Pipeline can be run using different values for all of the Runtime Inputs.

#### CI example

You can use Runtime Inputs in a CI stage's Infrastructure. Here's an example using a Runtime Input in the **Namespace** setting.

![](./static/runtime-inputs-05.png)

#### CD example

You can use Runtime Inputs for the Service in a CD stage's Service settings.

![](./static/runtime-inputs-06.png)

### Using runtime inputs during execution

Currently, this feature is behind the feature flag `NG_EXECUTION_INPUT`. Contact [Harness Support](mailto:support@harness.io) to enable the feature. 

You can add runtime input to a pipeline that runs when a stage or a step is executed. For example, for a custom stage with a Shell Script step that has an runtime input field, you will be prompted to enter the input during the execution just before starting the Shell Script step.

> If a custom stage is setup for runtime input, you can enter a shell script when prompted by Harness during execution. 

Reviewer 2: ```suggestion
If a custom stage is set up for runtime input, you can enter a shell script when prompted by Harness during execution. 
```

> If a Harness Approval step is setup for runtime input, when the pipeline executes, you can specify the Harness Groups that will approve that step.

Reviewer 2: ```suggestion
If a Harness Approval step is set up for runtime input, when the pipeline executes, you can specify the Harness Groups that will approve that step.
```

You can select the following runtime input types for a step or stage during an execution:  

* Allowed values (`allowedValues()`)  
* Default value (optional) (`default()`)
* Request input value when step/stage is being executed (`executionInput()`)

You can add execution inputs in this format: `<+input>.default(DefaultInputValue).executionInput()`.  
For example:
* `<+input>.allowedValues(value1,value2).executionInput()` - During execution, you will be prompted to enter `value 1` and `value 2` inputs. Only `value1` and `value2` values will be allowed as valid inputs.

> * `<+input>.allowedValues(value1,value2).default(value1).executionInput()` - During execution, you will be prompted to enter `value 1` and `value 2` inputs. Only `value1` and `value2` values will be allowed as valid inputs. `value1` is the default input in this example, so it appears as the deafult input in the prompt.  

Reviewer 2: ```suggestion
* `<+input>.allowedValues(value1,value2).default(value1).executionInput()` - During execution, you will be prompted to enter `value 1` and `value 2` inputs. Only `value1` and `value2` values will be allowed as valid inputs. `value1` is the default input in this example, so it appears as the default input in the prompt.  
```

You can also create default values in templates and later override them during the pipeline execution. 

> Here's a YAML example of a stage template with default value, `<+input>.default(“ABC”)`. You can ovveride this default value during pipeline execution. 

Reviewer 2: ```suggestion
Here's a YAML example of a stage template with default value, `<+input>.default(“ABC”)`. You can override this default value during pipeline execution. 
```

  orgIdentifier: harness
  tags: {}
  stages:
    - stage:
        name: s1
        identifier: s1
        template:
          templateRef: stageTempDefaults1
          versionLabel: v1
          templateInputs:
            type: Custom
            variables:
              - name: var1
                type: String
                default: ABC
                value: <+input>.default(ABC)
 ```

#### Limitations and requirements

The following limitations and requirements apply to this feature:

- A Harness user must have the **Pipeline Execute** permission to be able to submit runtime input during execution.

#### Using allowed values and default values in runtime inputs

You can set allowed values and a default value for a runtime input. The default value should be present in the list of allowed values. 

Multiple selection is allowed for runtime inputs defined for pipelines, stages, and shell script variables. 

The multiple selection functionality is currently behind the feature flag, `PIE_MULTISELECT_AND_COMMA_IN_ALLOWED_VALUES`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.

![](./static/runtime-inputs-11.png)

> Commas are supported in both allowed and default values. A string with comma must be wrapped within `\'   \'`. 

Reviewer 2: ```suggestion
Commas are supported in both allowed and default values. A string with a comma must be wrapped within `\'   \'`. 
```

Save and run the pipeline.

#### Using runtime input with an Approval step

> To add a runtime input with an Approval step, create a new pipeline, add an Approval stage, and click **Set Up Stage**.

Reviewer 2: ```suggestion
To add a runtime input with an Approval step, create a new pipeline, add an Approval stage, and then click **Set Up Stage**.
```
